extension/socks5proxy/src/socks5connection.cpp

```diff
@@ -254,7 +277,79 @@ qint64 Socks5Connection::proxy(QIODevice* a, QIODevice* b) {
   return bytes;
 }
 
-void Socks5Connection::configureOutSocket() {
+void Socks5Connection::dnsResolutionFinished(quint16 port) {
+  QDnsLookup* lookup = qobject_cast<QDnsLookup*>(QObject::sender());
+
+  // Garbage collect the lookup when we're finished.
+  m_dnsLookupAttempts--;
+  auto guard = qScopeGuard([lookup]() {
+    if (lookup->isFinished()) {
+      lookup->deleteLater();
```

Comment:
you delete the sender? Tell me why this is fine.

Reply:
We delete the sender if, and only if, the lookup is finished after exiting this function. This allows us to restart the lookup and continue DNS resolution simply by calling `lookup->lookup()` but if no such continuation is attempted then the `QDnsLookup` object gets garbage collected.

I suppose it is a little bit convoluted on how this works and prevents leaks or use-after-frees.

Reply:
For now, the only such case is when we receive a `CNAME` record and we need to restart the DNS lookup using the record pointed to by the `CNAME`